source/common/common/utility.cc

```diff
@@ -486,6 +486,33 @@ std::string StringUtil::removeCharacters(const absl::string_view& str,
   return absl::StrJoin(pieces, "");
 }
 
+bool StringUtil::consumeLeadingDigits(absl::string_view* s, uint64_t* val) {
+  const char* p = s->data();
+  const char* limit = p + s->size();
+  uint64_t v = 0;
+  while (p < limit) {
+    const char c = *p;
+    if (c < '0' || c > '9') {
+      break;
+    }
+    uint64_t new_v = (v * 10) + (c - '0');
```

Comment:
Speaking of v, how would you feel about replacing all uses of v with *val? I can see the advantage of not modifying val in case of overflow, but since that should be an exceptional case I don't think you need to tread too lightly in terms of not causing a disturbance.

Alternatively, regarding that overflow case, since we're in Envoy, it might be best to throw an exception in the case of an overflow. Then you could replace the val out-param with a nice simple return, and you could be confident that no user will ever ignore a `false` return and blithely proceed on with a bad value.

Reply:
I'm thinking the reason for not setting `*val` until we're confident we've parsed a valid string of digits is to communicate intent when reading the code... that we're not sure the value we're collecting along the way is actually worth something until we've hit a boundary (end of string or non-numeric character) and can confidently return what we've collected. If we accumulate in `*val` while processing, (and reset if we hit an invalid value), it seems to be a bit backwards to me, in that we are collecting a result before we can confidently assert that it should be collected.

That said, with both logic and maintainability, the two approaches are roughly equivalent, I just prefer only copying the value when we know for sure it's gonna stick.

Reply:
Sure, that makes sense.

I guess the real point worth discussing is what evolved into the 2nd paragraph (honestly I probably should have just deleted the 1st, haha). The current version is vulnerable to incorrect usage if a particularly careless user comes along. I suggested exceptions because Envoy makes me think exceptions, but other good (perhaps better) approaches would be returning a std::optional or absl::StatusOr.

Reply:
Envoy style slightly discourages exceptions on the data path. How about this:
optional<uint64_t> consumeLeadingDigits(string_view&)

(Note also the change from pointer to reference.)

Reply:
For simplicity, what do you think of first getting a string_view that's just the digits, then calling absl::SimpleAtoi?

Reply:
To capture the digits, would you use a regex or something else? This implementation is an order of magnitude faster than using a regex, but perhaps there's another approach I'm not thinking of?

Reply:
In eatLeadingDuration, I use c_find_if_not(s, &absl::ascii_isdigit).

https://github.com/toddmgreer/envoy/blob/f3161a256ff210c049f366fe0d02cc0b033767ce/source/extensions/filters/http/cache/http_cache_utils.cc#L90

In the above

Reply:
I modified the code from eatLeadingDuration to output an `int`. It seems reasonable:

```
bool StringUtil::consumeLeadingDigits2(absl::string_view* s, int* val) {                                                                                                                                                                                 
  const absl::string_view::iterator digits_end = absl::c_find_if_not(*s, &absl::ascii_isdigit);                                                                                                                                                          
  const size_t digits_length = digits_end - s->begin();                                                                                                                                                                                                  
  if (digits_length == 0) {                                                                                                                                                                                                                              
    return false;                                                                                                                                                                                                                                        
  }                                                                                                                                                                                                                                                      
  const absl::string_view digits(s->begin(), digits_length);                                                                                                                                                                                             
  s->remove_prefix(digits_length);                                                                                                                                                                                                                       
  return absl::SimpleAtoi(digits, val);                                                                                                                                                                                                                  
}  
```

Benchmarked, the current algorithm is ~6x faster, though I haven't dug into precisely why, just yet (I converted the current one to handle ints so the input/output were the same across functions):

```
#include <benchmark/benchmark.h>                                                                                                                                                                                                                          
#include "common/common/utility.h"                                                                                                                                                                                                                        
namespace Envoy {                                                                                                                                                                                                                                         
static void BM_ConsumeLeadingDigits(benchmark::State& state) {                                                                                                                                                                                            
  absl::string_view input{"1234567890xyz"};                                                                                                                                                                                                               
  int output;                                                                                                                                                                                                                                             
  for (auto _ : state) {                                                                                                                                                                                                                                  
    StringUtil::consumeLeadingDigits(&input, &output);                                                                                                                                                                                                    
  }                                                                                                                                                                                                                                                       
}                                                                                                                                                                                                                                                         
BENCHMARK(BM_ConsumeLeadingDigits);                                                                                                                                                                                                                       
static void BM_ConsumeLeadingDigits2(benchmark::State& state) {                                                                                                                                                                                           
  absl::string_view input{"1234567890xyz"};                                                                                                                                                                                                               
  int output;                                                                                                                                                                                                                                             
  for (auto _ : state) {                                                                                                                                                                                                                                  
    StringUtil::consumeLeadingDigits2(&input, &output);                                                                                                                                                                                                   
  }                                                                                                                                                                                                                                                       
}                                                                                                                                                                                                                                                         
BENCHMARK(BM_ConsumeLeadingDigits2);                                                                                                                                                                                                                      
}                                                                                                                                                                                                                                                         
BENCHMARK_MAIN();  
```


```
-------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations
-------------------------------------------------------------------
BM_ConsumeLeadingDigits        6.51 ns         6.51 ns     97609322
BM_ConsumeLeadingDigits2       40.8 ns         40.8 ns     17198831
```

Maybe both are fast enough that the difference doesn't matter?

Reply:
Alternately, perhaps we can use the faster version for both range headers and `eatLeadingDuration`

Reply:
I think both are fast enough. In both Envoy and GFE, the preference is generally to write code that's as clear (and as clearly corrrect) as possible (as long as it's not gratuitously slow/big), and to only optimize when something turns out to be too slow or too big.

consumeLeadingDigits2 is easier to understand, and easier to be convinced of its correctness--go with that.

Nits: pass by reference instead of pointer, and return optional<uint64_t>

Reply:
Regardless of which way you go with consumeLeadingDigits (per offline discussion, either way is fine), eatLeadingDuration should be changed to begin with a call to consumeLeadingDigits.

Reply:
The names of eatLeadingDuration and consumeLeadingDigits should start with the same verb. I went with 'eat' for the former because it has Envoy-style usage--I fear that calling it 'consume' would cause people to expect to call it like absl's 'consume' functions.

Reply:
I think I've resolved all the comments in this thread. Feel free to speak up if I missed any of the above.